src/writer/core.py

```diff
@@ -659,6 +678,48 @@ def _set_state_item(self, key: str, value: Any):
                 self._state_proxy[key] = value
 
 
+    def subscribe_mutation(self, path: Union[str, List[str]], handler: Callable[['State'], None], initial_triggered: bool = False) -> None:
+        """
+        Automatically triggers a handler when a mutation occurs in the state.
+
+        >>> def _increment_counter(state):
+        >>>     state_proxy['my_counter'] += 1
+        >>>
+        >>> state = WriterState({'a': 1, 'c': {'a': 1, 'b': 3}, 'my_counter': 0})
+        >>> state.subscribe_mutation('a', _increment_counter)
+        >>> state.subscribe_mutation('c.a', _increment_counter)
+        >>> state['a'] = 2 # will trigger _increment_counter
+        >>> state['a'] = 3 # will trigger _increment_counter
+        >>> state['c']['a'] = 2 # will trigger _increment_counter
+
+        :param path: path of mutation to monitor
+        :param func: handler to call when the path is mutated
+        """
+        if isinstance(path, str):
+            path_list = [path]
+        else:
+            path_list = path
+
+        for p in path_list:
+            state_proxy = self._state_proxy
+            path_parts = p.split(".")
```

Comment:
This doesn't handle key containing `.` .

Also, it doesn't work with Array index because the `path_parts` is an array of string (see my comment below)

IMO

- it should handle a path like this one `foo["bar.baz"][0]` we might need a library to handle it
- Or we should only accept array of string/number like `["foo", "bar.baz", 0]`
- Or we considerate it's OK, but it might need a comment in the doc saying this limitation

Reply:
**usage of `.` into a state key trigger errors**. That's a good point, we don't manage properly the case where a state contains `.` in the key. Currently it raise many issue. 

I would advocate to manage the character `.`  as `reserved keyword` for State key. We should ignore this value and show warning when the application is loaded.

![image](https://github.com/user-attachments/assets/0a74adf7-cf85-45a1-840c-f75f831f5e93)

**mutation of State are not monitored on list** : the `State` structure on the backend side supervises mutations exclusively on dictionaries, not on lists. This is an intrinsic limitation of WF.

A developer has to reassign a list to trigger the mutation. This point is covered in the [Mutation detection](https://dev.writer.com/framework/event-handlers#mutation-detection) in doc. It is therefore not possible to subscribe to a mutation on an element of a list.

Reply:
Clear, thanks for the explanations.

About array, is it something we want to support ? Use case I see is to get the sum of elements in an array to get the sum. Something like this

```python
class MyState(wf.WriterState):
  arr: List[int]
  total: int

def cumulative_sum(state: MyState):
  state.total = sum(state.arr)

initial_state = ss.init_state({
    "arr": [1,2,3,4],
    "total": 0
}, schema=MyState)

initial_state.subscribe_mutation('counter', cumulative_sum)

initial_state['counter'][0] = 9
```

Reply:
I thin this question has already been debated in the past. I don't find trace about that. I think we decided that for the moment we wanted to avoid this pattern. Re-encoding a list as an observable object would have an overhead that we want to avoid. 

In Python, lists are more often reconstructed rather than modified.

However, this was before the existence of typed states. Typed states allow you to specify how to behave with respect to a collection. I wonder on which usecase we would benefit from implementing this pattern.